test/EngineTest.cpp

```diff
@@ -43,94 +57,259 @@ IdTable makeIdTableFromVector(std::vector<std::array<size_t, TABLE_WIDTH>> table
+ *  together. Look into src/engine/Join.h for how it should look like.
+ *
+ * @returns tableA and tableB joined together in a IdTable.
+ */
+template<size_t TABLE_A_WIDTH, size_t TABLE_B_WIDTH, typename JOIN_FUNCTION>
+IdTable useJoinFunctionOnVectorsTables(
+        const std::vector<std::array<size_t, TABLE_A_WIDTH>>& tableA,
+        const size_t jcA,
+        const std::vector<std::array<size_t, TABLE_B_WIDTH>>& tableB,
+        const size_t jcB,
+        JOIN_FUNCTION func,
+        ad_utility::source_location T = ad_utility::source_location::current()
+      ) {
+  // For generating better messages, when failing a test.
+  auto trace = generateLocationTrace(T, "useJoinFunctionOnVectorsTables");
+
+  IdTable a = makeIdTableFromVector(tableA);
+  IdTable b = makeIdTableFromVector(tableB);
+  
+  constexpr int reswidth = TABLE_A_WIDTH + TABLE_B_WIDTH - 1;
+  IdTable res(reswidth, allocator());
+  
+  CALL_FIXED_SIZE_3(TABLE_A_WIDTH, TABLE_B_WIDTH, reswidth, func.template operator(), a, jcA, b, jcB, &res);
+
+  return res;
+}
 
-  for (size_t i = 1; i <= 10000; ++i) {
-    a.push_back({I(4 + i), I(2 + i)});
+/*
+ * @brief Goes through the sets of tests, converts the given vectors to
+ * IdTables, joins them together with the given join function and
+ * compares the results with the given sample solution.
+ * However, because of technical limitations, all left/right/sample-solution
+ * vectors MUST have the same amount of columns.
+ *
+ * @tparam TABLE_A_WIDTH is the width of the left tables.
+ * @tparam TABLE_B_WIDTH is the width of the right tables.
+ * @tparam TABLE_RESULT_WIDTH is the width of the sample solution tables.
+ * @tparam How many test cases we have.
+ * @tparam JOIN_FUNCTION is used to allow the transfer of any type of
+ *  lambda function, that could contain a join function. You never have to
+ *  actually specify this parameter, just let inference do its job.
+ *
+ * @param testSet is an array of tests, wrritten as tuples. The items in
+ * the tuple have the following meaning:
+ *  - First item: The left IdTable described through a vector of arrays.
+ *  - Second item: The join column of the left IdTable.
+ *  - Third item: The right IdTable described through a vector of arrays.
+ *  - Fourth item: The join column of the right IdTable.
+ *  - Fith item: The sample solution of the join function performed on both
+ *    described IdTables. Once again, this IdTable is described through a
+ *    vector of arrays.
+ * @param func the function, that will be used for joining the two tables
+ *  together. Look into src/engine/Join.h, or this file, for how it should look like.
+ * @param testForSorted If this is true, then the result of func on the entries
+ *  of testSet will also be tested on if they are sorted by their join column.
+ */
+template<size_t TABLE_A_WIDTH, size_t TABLE_B_WIDTH, size_t TABLE_RESULT_WIDTH, size_t NUMBER_OF_TESTS, typename JOIN_FUNCTION>
+void goThroughSetOfTestsWithJoinFunction(
```

Comment:
At this place (before a further iteration),  I suggest some simplifications of boiler plate code:

- You can always use `makeIdTableFromVector<2>({{1, 2}, {1, 3}})` instead of `makeIdTableFromVector(std::vector<std::array<...`.
- you should definitely at the beginning of the file somewhere create an alias for the  `vector<array<size_t, N>>`, 
  for example 
```
template<size_t N>
using Table = std::vector<std::array<size_t, N>>;
```
then you can write
`Table<2> table {{1,2}, {2, 3}};` instead of `vector<array<size_t, 2> table = ....`. Which will make a lot of code much more readable.

- Your test case shouldn't be a tuple but a proper struct `JoinTestCase` that is templated on the three (two?) widths.
- Isn't the output width here ALWAYS `LEFT_WIDTH + RIGHT_WIDTH - 1`, thus you could get rid of a lot of third template parameters.

- The outer data structure of the variable `testSet` here can be a `vector` instead of `array`, then you get rid of the `NUMBER_OF_TESTS`.


With all those changes, the signature of this function becomes
``` 
template<size_t A_WIDTH, size_T B_WIDTH, typename JOIN_FUNCTION>
void goThrough....(std::vector<TestCase<A_WIDTH, B_WIDTH>> testSet, JOIN_FUNCTION joinFunction, ...)
```
Which I personally consider much less intimidating.

Reply:
Most of this was already done, with the exception of an alias for an often used type, which I did in [273987f](https://github.com/ad-freiburg/qlever/pull/821/commits/273987f0186b66bc4594881d1f888dfc789309d5).